### PR TITLE
Implemented TensorVector bindings and `sm.concatenate`

### PR DESCRIPTION
**Pull request by yushiyangk**
For each TensorVector, a Typescript `Array<Tensor>` is converted to a `BigInt64Array` containing pointers to `Tensor`s. This is passed to the FFI as `tensors_ptr` and `tensors_len` which have types `FFIType.ptr` and `FFIType.i64`. In C++, we cast each `uint64_t` in the array to `fl::Tensor*` and dereference it to obtain `std::vector<fl::Tensor>`, which is the type Flashlight expects.

### Gradient dependencies

TensorVector is unfolded into individual Tensors before being passed to gradient calculations. As a consequence of this, in general each function can only take at most one TensorVector parameter; if there were multiple TensorVector parameters, the gradient function would be unable to distinguish which TensorVector a given Tensor came from (without some horrible hacks). The alternative would have been to pass TensorVectors as-is, but I think that would require modifying the `Grad` interface to include a "sub-index" field, which I'm not sure is desirable.

Changed `Grad.in` interface from `Tensor[]` to `any[]`, which it already was de facto since `Tensor.deps` contained non-Tensor parameters as well. Calculating the gradient of `sm.concatenate` is much more straightforward when the `axis` parameter is available.

Since these `Tensor.deps` are being used within Typescript itself, and not being passed to the FFI, I changed the codegen to keep them as Typescript-native types. Previously, they were stored as the same ptr, len pairs that were passed to the FFI.

### `sm.concatenate`

Flashlight's concatenate seems to support automatic expanding the dimensionality of resultant Tensors when a negative `axis` is passed. This functionality is preserved in the Typescript bindings, and reflected in the tests. However, Flashlight behaves inconsistently when a 0-dimensional Tensors are passed with a negative `axis`, always returning an unconcatenated 0-dimensional Tensor instead of expanding its dimensionality and concatenating it. I added a bit of logic to make `sm.concatenate` treat scalar Tensors consistently as well.

### `Tensor.index`

Changed this method to take ranges in the form of `'0:3'` to take a slice of the Tensor from the 0th to the 2nd index (inclusive) in that particular dimension. E.g.: `sm.full([2, 3, 4], 1).index([1, ':', '0:2'])`.

This was needed to get the gradient of `sm.concatenate`.

### CODE REVIEW

File: scripts/gen_binding.py [hunk 1]
```diff
@@ -444,16 +481,16 @@ def normalize_arg(arg):
...
-        wrap_args = [('this', 'tensor')] + wrap_args
-    js_ptr_args = [x[0] if x[1] != "tensor" else f"{x[0]}.ptr" for x in wrap_args]
+        wrap_args = [('this', 'Tensor')] + wrap_args
+        js_tensor_args = ["this"] + js_tensor_args
+    js_ptr_args = [x[0] if x[1] != "Tensor" else f"{x[0]}.ptr" for x in wrap_args]
     js_ptr_result = f"const _ptr = fl._{op}({', '.join(js_ptr_args)})"
-    js_tensor_args = [f"{arg[0]}" for arg in wrap_args if arg[1] == "tensor"]
-    js_provenance_args = '||'.join([f"{t}.provenance" for t in js_tensor_args])
-    js_requires_grad_args = '||'.join([f"{t}.requires_grad" for t in js_tensor_args])
+    js_provenance_args = '||'.join([f"{t}.provenance" for t in js_tensor_args] + [f"{tv}.some((t) => t.provenance)" for tv in js_tensor_vector_args])
```
Comment (bwasti) [inline scripts/gen_binding.py:488]: provenance actually contains information about where the tensor is from (non-zero hash of root device and populated during network transfer).  The original logic preserves the information by short circuiting, whereas `some` generates a boolean.

The `requires_*` are logically sound, but I believe early benchmarking showed the JSC JIT was able to handle explicit chained `||` faster than lambda evaluation.  I know it's ugly, but it's the only way I was able to get performance (this code is pretty hot).

Please revert these bits :)

File: shumai/ffi/ffi_bind_utils.ts [hunk 2]
```diff
@@ -1,25 +1,43 @@
...
+        throw new Error(`Unsupported type at position ${i} of input arg`)
+      }
+    }
+  } else {
+    numArray = arg
+  }
+
+  let outArray: Int32Array | BigInt64Array
   switch (type) {
     case FFIType.i32:
```
Comment (bwasti) [inline shumai/ffi/ffi_bind_utils.ts:30]: let's just kill this (i32 arrays) with 🔥 for now :)

File: shumai/tensor/register_gradients.ts [hunk 3]
```diff
@@ -5,11 +5,19 @@ const sm = { ...base, ...ops }
 
 export interface Grad {
   idx: number
-  in: Tensor[]
+  in: any[]
   grad_in: Tensor
   out: Tensor
 }
 
+function getBackwardGradient(grad: Grad): Tensor {
```
Comment (bwasti) [inline shumai/tensor/register_gradients.ts:13]: nit: is this going to be used outside `concatenate`'s gradient? may be worth inlining?

File: scripts/gen_binding.py [hunk 4]
```diff
@@ -444,16 +481,16 @@ def normalize_arg(arg):
...
+    js_requires_grad_args = '||'.join([f"{t}.requires_grad" for t in js_tensor_args] + [f"{tv}.some((t) => t.requires_grad)" for tv in js_tensor_vector_args])
     js_requires_grad_args = 'false' if len(js_requires_grad_args) == 0 else js_requires_grad_args
-    js_requires_stats = '||'.join([f"{t}.requires_stats" for t in js_tensor_args])
+    js_requires_stats = '||'.join([f"{t}.requires_stats" for t in js_tensor_args] + [f"{tv}.some((t) => t.requires_stats)" for tv in js_tensor_vector_args])
     js_requires_stats = 'false' if len(js_requires_stats) == 0 else js_requires_stats
-    js_grad_args = (['this'] if methods_only else []) + js_args
+    js_grad_args = (['this'] if methods_only else []) + [f"...{n}" if t == "TensorVector" else f"{n}" for n, t in zip(js_grad_args, js_grad_arg_types)]
```
Comment (bwasti) [inline scripts/gen_binding.py:493]: I believe the logic previously was actually wrong.  `js_grad_args` should only be used for `js_deps` (to traverse backward) - which should really just be `Tensor`s.  Right now it includes `Shape`, `Axes` and other types.

It's not breaking, but I think we can actually clean up the differentiation code later if we can safely assume every dep is in fact a tensor.

**Review by bwasti — CHANGES_REQUESTED** [review]
This is great! thanks for making all these changes.

Only one required change and some nits.

File: scripts/gen_binding.py (hunk 4 above)
Comment (yushiyangk) [reply scripts/gen_binding.py:493]: It seems to me that having the other types of arguments makes calculating gradient more straightforward, at least in the case of `concatenate`. Without it, we would have to add a bit of logic to infer the axis of concatenation based on the shapes of the input and output.

File: scripts/gen_binding.py (hunk 1 above)
Comment (yushiyangk) [reply scripts/gen_binding.py:488]: I don't think we can generate explicit `||`s for the members of a TensorVector at codegen time since we don't know the length of the vector.

Is a for loop more performant than lambdas? I should be able to use a loop to preserve the return type of provenance as well.

File: scripts/gen_binding.py (hunk 4 above)
Comment (bwasti) [reply scripts/gen_binding.py:493]: ah you're totally right - I thought we were using a different way of passing inputs here: https://github.com/facebookresearch/shumai/blame/main/shumai/tensor/tensor.ts#L110

please disregard this comment :)

File: scripts/gen_binding.py (hunk 1 above)
Comment (bwasti) [reply scripts/gen_binding.py:488]: hmm good point on generating explicit `||` for TensorVector - we can use `tensor_vector.reduce((x,y) => x || y)` and leave the other args as is

**Comment by bwasti** [comment]
awesome - all the tests pass so I'm landing :)
